Refactor OpenSearchSchema to be more robust.

mdwelsh opened the pull request:
Should make retrieval of the schema more robust when there are mapping fields not present in the examples.

baitsguy commented on lib/sycamore/sycamore/query/schema.py:
```diff
@@ -21,15 +21,16 @@ def get_schema(self) -> dict[str, str]:
-                result[key] = f"({type(random_sample[0]['_source']['properties']['entity'][k[18:]])}) e.g. "
+                try:
+                    key = k
+                    result[key] = f"({type(random_sample[0]['_source']['properties']['entity'][k[18:]])}) e.g. "
```
this will break the formatting if the first "random_sample" doesn't have the key I think?

tranade commented on lib/sycamore/sycamore/query/schema.py:
```diff
@@ -21,15 +21,16 @@ def get_schema(self) -> dict[str, str]:
 
         for k, v in schema[self._index]["mappings"].items():
             if k.startswith("properties.entity") and ".keyword" not in k:
```
should we also add a check for if k has a "." in it? this could reduce the # of exceptions we have to catch

mdwelsh commented:
I reworked this PR so that it returns results in a more structured (and hence type-checkable) way, and defers how the schema is rendered into the LLM prompt to the planner code. Also added a test.

baitsguy commented on lib/sycamore/sycamore/tests/unit/query/test_schema.py:
```diff
@@ -0,0 +1,67 @@
+                    {
+                        "_source": {
+                            "properties": {
+                                "entity": {
```
maybe have an entry that is missing a property

mdwelsh replied:
Done.

mdwelsh commented:
OK, updated to support missing properties in some docs.

tranade commented on lib/sycamore/sycamore/query/planner.py:
```diff
@@ -415,6 +427,7 @@ def generate_prompt(self, query):
         prompt += f"""
         USER QUESTION: {query}
         """
+        print(prompt)
```
do we want this here?

mdwelsh replied:
Oops, good catch! Thanks!
